File: firmware/src/services/ble/analysis.rs
use crate::services::measure::Measurement;

use super::model::PROTOCOL_VERSION;
use super::{Alcohol, DeviceEvent, MeasurementResult, PhoneContext, Pulse, Risk};

const LEGAL_LIMIT_MILLI_PERCENT: u16 = 30;
const CAUTION_MILLI_PERCENT: u16 = 15;
const DEFAULT_BAC_ELIMINATION_MILLI_PERCENT_PER_HOUR: u16 = 15;
const DEFAULT_BRAC_NOISE_MG_L_X1000: u16 = 10;
const MIN_CONTEXT_ELIMINATION_MG_L_PER_HOUR_X1000: u16 = 20;
const MAX_CONTEXT_ELIMINATION_MG_L_PER_HOUR_X1000: u16 = 120;
const MAX_SOBER_BASELINE_MG_L_X1000: u16 = 50;
const MAX_SOBER_BASELINE_MAD_MG_L_X1000: u16 = 50;

pub fn measurement_result(
    session_id: String,
    measurement: Measurement,
    context: Option<&PhoneContext>,
) -> DeviceEvent {
    let alcohol_mg_l_x1000 = measurement.alcohol_mg_l_x1000();
    let pulse_bpm = measurement.pulse_bpm();
    let corrected_alcohol = corrected_alcohol_mg_l_x1000(alcohol_mg_l_x1000, context);
    let upper_alcohol = upper_alcohol_mg_l_x1000(alcohol_mg_l_x1000, context);
    let raw_bac_milli_percent = brac_to_bac_milli_percent(alcohol_mg_l_x1000);
    let bac_milli_percent = brac_to_bac_milli_percent(corrected_alcohol);
    let bac_upper_milli_percent =
        brac_to_bac_milli_percent(upper_alcohol).max(raw_bac_milli_percent);
    let risk = risk_from_upper_bac(bac_upper_milli_percent);
    let confidence_percent = confidence_percent(context, pulse_bpm);

    DeviceEvent::MeasurementResult(MeasurementResult {
        v: PROTOCOL_VERSION,
        session_id,
        measured_at_unix_ms: context.and_then(|context| context.phone_time_unix_ms),
        alcohol: Alcohol {
            mg_l_x1000: alcohol_mg_l_x1000,
        },
        pulse: Some(Pulse {
            bpm: pulse_bpm as f32,
            stable: true,
            confidence_percent: pulse_confidence_percent(context, pulse_bpm),
        }),
        bac_milli_percent: Some(bac_milli_percent),
        bac_upper_milli_percent: Some(bac_upper_milli_percent),
        sober_time_minutes: estimate_sober_time_minutes(
            upper_alcohol,
            bac_upper_milli_percent,
            context,
        ),
        risk,
        confidence_percent,
    })
}

fn corrected_alcohol_mg_l_x1000(alcohol_mg_l_x1000: u16, context: Option<&PhoneContext>) -> u16 {
    let baseline = context.and_then(sober_baseline_mg_l_x1000).unwrap_or(0);

    alcohol_mg_l_x1000.saturating_sub(baseline)
}

fn upper_alcohol_mg_l_x1000(alcohol_mg_l_x1000: u16, context: Option<&PhoneContext>) -> u16 {
    let Some(context) = context else {
        return alcohol_mg_l_x1000.saturating_add(DEFAULT_BRAC_NOISE_MG_L_X1000);
    };

    let baseline = sober_baseline_mg_l_x1000(context).unwrap_or(0);
    let baseline_noise = sober_baseline_mad_mg_l_x1000(context)
        .map(|mad| mad.saturating_mul(3))
        .unwrap_or(DEFAULT_BRAC_NOISE_MG_L_X1000)
        .max(DEFAULT_BRAC_NOISE_MG_L_X1000);

    alcohol_mg_l_x1000.saturating_sub(baseline.saturating_sub(baseline_noise))
}

fn sober_baseline_mg_l_x1000(context: &PhoneContext) -> Option<u16> {
    context
        .sober_alcohol_mg_l_x1000
        .filter(|baseline| *baseline <= MAX_SOBER_BASELINE_MG_L_X1000)
}

fn sober_baseline_mad_mg_l_x1000(context: &PhoneContext) -> Option<u16> {
    context
        .sober_alcohol_mad_mg_l_x1000
        .map(|mad| mad.min(MAX_SOBER_BASELINE_MAD_MG_L_X1000))
}

fn brac_to_bac_milli_percent(alcohol_mg_l_x1000: u16) -> u16 {
    ((u32::from(alcohol_mg_l_x1000) * 21 + 50) / 100).min(u32::from(u16::MAX)) as u16
}

fn estimate_sober_time_minutes(
    upper_alcohol_mg_l_x1000: u16,
    bac_upper_milli_percent: u16,
    context: Option<&PhoneContext>,
) -> Option<u16> {
    if upper_alcohol_mg_l_x1000 == 0 || bac_upper_milli_percent < CAUTION_MILLI_PERCENT {
        return Some(0);
    }

    if let Some(rate) = context.and_then(|context| context.elimination_mg_l_per_hour_x1000) {
        if (MIN_CONTEXT_ELIMINATION_MG_L_PER_HOUR_X1000
            ..=MAX_CONTEXT_ELIMINATION_MG_L_PER_HOUR_X1000)
            .contains(&rate)
        {
            return Some(minutes_for_rate(upper_alcohol_mg_l_x1000, rate));
        }
    }

    Some(minutes_for_rate(
        bac_upper_milli_percent,
        DEFAULT_BAC_ELIMINATION_MILLI_PERCENT_PER_HOUR,
    ))
}

fn minutes_for_rate(amount: u16, rate_per_hour: u16) -> u16 {
    if rate_per_hour == 0 {
        return u16::MAX;
    }

    ((u32::from(amount) * 60 + u32::from(rate_per_hour) - 1) / u32::from(rate_per_hour))
        .min(u32::from(u16::MAX)) as u16
}

fn risk_from_upper_bac(bac_upper_milli_percent: u16) -> Risk {
    if bac_upper_milli_percent >= LEGAL_LIMIT_MILLI_PERCENT {
        Risk::Danger
    } else if bac_upper_milli_percent >= CAUTION_MILLI_PERCENT {
        Risk::Caution
    } else {
        Risk::Safe
    }
}

fn confidence_percent(context: Option<&PhoneContext>, pulse_bpm: u16) -> u8 {
    let mut confidence = 55_u8;

    if let Some(context) = context {
        if sober_baseline_mg_l_x1000(context).is_some() {
            confidence = confidence.saturating_add(10);
        }

        if sober_baseline_mad_mg_l_x1000(context).is_some() {
            confidence = confidence.saturating_add(5);
        }

        if context.elimination_mg_l_per_hour_x1000.is_some() {
            confidence = confidence.saturating_add(10);
        }

        if !context.recent.is_empty() {
            confidence = confidence.saturating_add(5);
        }

        if let Some(resting_bpm) = context.resting_bpm {
            let delta = pulse_bpm.abs_diff(resting_bpm);

            if delta <= 20 {
                confidence = confidence.saturating_add(5);
            } else {
                confidence = confidence.saturating_sub(10);
            }
        }
    }

    confidence.min(90)
}

fn pulse_confidence_percent(context: Option<&PhoneContext>, pulse_bpm: u16) -> u8 {
    let Some(resting_bpm) = context.and_then(|context| context.resting_bpm) else {
        return 65;
    };

    let delta = pulse_bpm.abs_diff(resting_bpm);

    if delta <= 15 {
        85
    } else if delta <= 30 {
        70
    } else {
        45
    }
}

File: app/src/lib/personalization/analysis.ts

import type { MeasurementResult } from '@/lib/ble/model';
import type { UserBaseline } from '@/lib/storage/profile';

const legalLimitMilliPercent = 30;
const cautionMilliPercent = 15;
const defaultBracNoiseMgLX1000 = 10;
const minEliminationMgLPerHourX1000 = 20;
const maxEliminationMgLPerHourX1000 = 120;
const maxSoberBaselineMgLX1000 = 50;
const maxSoberBaselineMadMgLX1000 = 50;

/** raw 측정값과 앱 로컬 상태로 보수적인 운전 위험 분석을 계산한다. */
export function analyzeMeasurement(
  result: MeasurementResult,
  state: MeasurementAnalysisState
): MeasurementAnalysis {
  const alcohol = result.alcohol_mg_l_x1000;
  const baseline = soberBaselineMgLX1000(state.baseline) ?? 0;
  const correctedAlcohol = Math.max(0, alcohol - baseline);
  const upperAlcohol = upperAlcoholMgLX1000(alcohol, state.baseline);
  const rawBac = bracToBacMilliPercent(alcohol);
  const bacMilliPercent = bracToBacMilliPercent(correctedAlcohol);
  const bacUpperMilliPercent = Math.max(bracToBacMilliPercent(upperAlcohol), rawBac);

  return {
    bac_milli_percent: bacMilliPercent,
    bac_upper_milli_percent: bacUpperMilliPercent,
    risk: riskFromUpperBac(bacUpperMilliPercent),
    sober_time_minutes: soberTimeMinutes(
      upperAlcohol,
      bacUpperMilliPercent,
      state.baseline.elimination_mg_l_per_hour_x1000,
      state.conservativeEliminationMgLPerHourX1000
    ),
    confidence_percent: confidencePercent(result, state),
  };
}

function upperAlcoholMgLX1000(alcohol: number, baseline: UserBaseline) {
  const soberBaseline = soberBaselineMgLX1000(baseline) ?? 0;
  const soberBaselineMad = soberBaselineMadMgLX1000(baseline);
  const baselineNoise = Math.max(
    defaultBracNoiseMgLX1000,
    soberBaselineMad === null ? defaultBracNoiseMgLX1000 : soberBaselineMad * 3
  );

  return Math.max(0, alcohol - Math.max(0, soberBaseline - baselineNoise));
}

function soberBaselineMgLX1000(baseline: UserBaseline) {
  const value = baseline.sober_alcohol_mg_l_x1000;
  return value !== null && value <= maxSoberBaselineMgLX1000 ? value : null;
}

function soberBaselineMadMgLX1000(baseline: UserBaseline) {
  const value = baseline.sober_alcohol_mad_mg_l_x1000;
  return value === null ? null : Math.min(value, maxSoberBaselineMadMgLX1000);
}

export function bracToBacMilliPercent(alcoholMgLX1000: number) {
  return Math.min(65535, Math.floor((alcoholMgLX1000 * 21 + 50) / 100));
}

function soberTimeMinutes(
  upperAlcoholMgLX1000: number,
  bacUpperMilliPercent: number,
  storedEliminationMgLPerHourX1000: number | null,
  conservativeEliminationMgLPerHourX1000: number
) {
  if (upperAlcoholMgLX1000 === 0 || bacUpperMilliPercent < cautionMilliPercent) {
    return 0;
  }

  const eliminationMgLPerHourX1000 = validElimination(storedEliminationMgLPerHourX1000)
    ? storedEliminationMgLPerHourX1000
    : conservativeEliminationMgLPerHourX1000;

  return Math.min(65535, Math.ceil((upperAlcoholMgLX1000 * 60) / eliminationMgLPerHourX1000));
}

function validElimination(value: number | null): value is number {
  return (
    value !== null &&
    value >= minEliminationMgLPerHourX1000 &&
    value <= maxEliminationMgLPerHourX1000
  );
}

// @ts-ignore
function riskFromUpperBac(bacUpperMilliPercent: number): Risk {
  if (bacUpperMilliPercent >= legalLimitMilliPercent) {
    return 'danger';
  }

  if (bacUpperMilliPercent >= cautionMilliPercent) {
    return 'caution';
  }

  return 'safe';
}

function confidencePercent(result: MeasurementResult, state: MeasurementAnalysisState) {
  let confidence = 55;

  if (soberBaselineMgLX1000(state.baseline) !== null) {
    confidence += 10;
  }

  if (soberBaselineMadMgLX1000(state.baseline) !== null) {
    confidence += 5;
  }

  if (state.baseline.elimination_mg_l_per_hour_x1000 !== null) {
    confidence += 10;
  }

  if (state.recentMeasurementCount > 0) {
    confidence += 5;
  }

  const restingBpm = state.baseline.resting_bpm;

  if (restingBpm !== null && result.pulse.status === 'measured') {
    confidence += Math.abs(result.pulse.bpm - restingBpm) <= 20 ? 5 : -10;
  } else if (restingBpm !== null) {
    confidence -= 5;
  }

  return Math.min(90, confidence);
}

export interface MeasurementAnalysisState {
  baseline: UserBaseline;
  conservativeEliminationMgLPerHourX1000: number;
  recentMeasurementCount: number;
}

export interface MeasurementAnalysis {
  bac_milli_percent: number;
  bac_upper_milli_percent: number;
  risk: Risk;
  sober_time_minutes: number;
  confidence_percent: number;
}

export type Risk = 'safe' | 'caution' | 'danger';
